pkg/net/bitswap.go

```diff
@@ -151,3 +163,159 @@ func (c BitswapClient) Retrieve(
+			}
+			if len(links) == 1 {
+				// Generally this should not happen as the root node should contain at least one AggregateManifest and other links
+				// however, there may be a different construction in the future where this is still valid, so we will attempt to retrieve
+				logger.Info("starting node contains only 1 link - will still attempt retrieval test")
+				nextIndex = 0
+			} else {
+				// To be safe, never grab the first link off the root as it may refer to the AggregateManifest
+				nextIndex = 1 + rand.Intn(len(links)-1)
+			}
+		} else {
+			// Logic for all other non-root nodes
+			if len(links) < 1 {
+				return task.NewErrorRetrievalResult(task.CannotTraverse, errors.Errorf("node at depth %d contains no links, cannot traverse any further", i)), nil
+			}
+			// randomly pick a link to go down
+			nextIndex = rand.Intn(len(links))
+		}
+
+		cidToRetrieve, err = cid.Parse(links[nextIndex].String())
+		if err != nil {
+			return task.NewErrorRetrievalResultWithErrorResolution(task.CIDCodecNotSupported, err), nil
+		}
+
+		i++ // To the next layer of the tree
+	}
+}
+
+// Returns the raw block data, the links, and error if any
+// Takes in `network` and `bswap` client, so that it can be used in a loop for traversals / multiple retrievals
+func (c BitswapClient) RetrieveBlock(
+	parent context.Context,
+	target peer.AddrInfo,
+	network bsnet.BitSwapNetwork,
+	bswap *bsclient.Client,
+	cid cid.Cid) (blocks.Block, error) {
+	logger := logging.Logger("bitswap_retrieve_block").With("cid", cid).With("target", target)
+
+	notFound := make(chan struct{})
+	network.Start(MessageReceiver{BSClient: bswap, MessageHandler: func(
+		ctx context.Context, sender peer.ID, incoming bsmsg.BitSwapMessage) {
+		if sender == target.ID && slices.Contains(incoming.DontHaves(), cid) {
+			logger.Info("Block not found")
+			close(notFound)
+		}
+	}})
+	connectContext, cancel := context.WithTimeout(parent, c.timeout)
+	defer cancel()
+	logger.Info("Connecting to target peer...")
+	err := c.host.Connect(connectContext, target)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to connect to target peer")
+	}
+
+	resultChan := make(chan blocks.Block)
+	errChan := make(chan error)
+	go func() {
+		logger.Debug("Retrieving block...")
+		blk, err := bswap.GetBlock(connectContext, cid)
+		if err != nil {
+			logger.Info(err)
+			errChan <- err
+		} else {
+			resultChan <- blk
+		}
+	}()
+	select {
+	case <-notFound:
+		return nil, errors.New("DONT_HAVE received from the target peer")
+
+	case blk := <-resultChan:
+		return blk, nil
+
+	case err := <-errChan:
+		return nil, errors.Wrap(err, "error received %s")
+	}
+}
+
+// Attempts to decode the block data into a node and return its links
+func FindLinks(ctx context.Context, blk blocks.Block) ([]datamodel.Link, error) {
+	if blk.Cid().Prefix().Codec == cid.Raw {
+		// Note: this case will happen at the bottom of the tree
+		return nil, errors.New("raw block encountered " + blk.Cid().String())
```

Comment:
Shouldn't this simply be returtning an empty list of links instead of an error?

Reply:
good catch, will change